src/compile/selection/transforms/legends.ts

```diff
@@ -0,0 +1,129 @@
+import {selector as parseSelector} from 'vega-event-selector';
+import {TransformCompiler} from './transforms';
+import {UnitModel} from '../../unit';
+import {NonPositionScaleChannel} from '../../../channel';
+import {LegendComponent} from '../../legend/component';
+import {forEachSelection, SelectionComponent, TUPLE} from '..';
+import {array, isString} from 'vega-util';
+import {Stream, MergedStream} from 'vega';
+import {SELECTION_ID, isLegendBinding, isLegendStreamBinding} from '../../../selection';
+import * as log from '../../../log';
+import {duplicate, varName} from '../../../util';
+import {TUPLE_FIELDS} from './project';
+import {TOGGLE} from './toggle';
+
+const legendBindings: TransformCompiler = {
+  has: (selCmpt: SelectionComponent<'single' | 'multi'>) => {
+    const spec = selCmpt.resolve === 'global' && selCmpt.bind && isLegendBinding(selCmpt.bind);
+    const projLen = selCmpt.project.items.length === 1 && selCmpt.project.items[0].field !== SELECTION_ID;
+    if (spec && !projLen) {
+      log.warn(log.message.LEGEND_BINDINGS_PROJECT_LENGTH);
+    }
+
+    return spec && projLen;
+  },
+
+  parse: (model, selCmpt, selDef, origDef) => {
+    // Binding a selection to a legend disables default direct manipulation interaction.
+    // A user can choose to re-enable it by explicitly specifying triggering input events.
+    if (!origDef.on) delete selCmpt.events;
+    if (!origDef.clear) delete selCmpt.clear;
+
+    if (origDef.on || origDef.clear) {
+      const legendFilter = 'event.item && indexof(event.item.mark.role, "legend") < 0';
+      for (const evt of selCmpt.events) {
+        evt.filter = array(evt.filter || []);
+        if (evt.filter.indexOf(legendFilter) < 0) {
+          evt.filter.push(legendFilter);
+        }
+      }
+    }
+
+    const evt = isLegendStreamBinding(selCmpt.bind) ? selCmpt.bind.legend : 'click';
+    const stream: Stream[] = isString(evt) ? parseSelector(evt, 'view') : array(evt);
+    selCmpt.bind = {legend: {merge: stream}};
+  },
+
+  topLevelSignals: (model, selCmpt: SelectionComponent<'single' | 'multi'>, signals) => {
+    const selName = selCmpt.name;
+    const stream = isLegendStreamBinding(selCmpt.bind) && (selCmpt.bind.legend as MergedStream);
+    const markName = (name: string) => (s: Stream) => {
+      const ds = duplicate(s);
+      ds.markname = name;
+      return ds;
+    };
+
+    for (const proj of selCmpt.project.items) {
+      if (!proj.hasLegend) continue;
+      const prefix = `${proj.field}_legend`;
+      const sgName = `${selName}_${prefix}`;
+      const hasSignal = signals.filter(s => s.name === sgName);
+
+      if (!hasSignal.length) {
+        const events = stream.merge
+          .map(markName(`${prefix}_symbols`))
```

Comment:
The prefix can be undefined for legends that show counts.

Reply:
Selections are not supported for count/aggregate fields, so with https://github.com/vega/vega-lite/pull/5305/commits/d05d65126d851e97458736c76600ea86d07db605 we warn and skip these types of definitions. Moreover, as of https://github.com/vega/vega-lite/pull/5305/commits/e6917e25c6ecb4c292e497bb5ef9bfd3db1a6758, we only output legend names/interactive flags if a corresponding selection is found. Together, these two commits should address this point.